### PR TITLE
Refactor Scoring into 2 different Scoring callbacks.

### PR DESCRIPTION
**Pull request by benjamin-work**
There is now a scoring callback BatchScoring that determines the score
after each batch and then averages (similar to previous Scoring), and
a new EpochScoring, that only determines scores after each epoch (more
correct values).

These callbacks share a lot of common code, which is why there is a
ScoringBase now. Tests were also refactored, should be much more
transparent now since we don't mock the fit loop anymore (formerly
in conftest.py).

We use some sklearn internals to make this work, which revealed some
problems (passing initialized dataset has unexpected behavior).

Addresses #55

### CODE REVIEW

File: skorch/tests/test_callbacks.py [hunk 1]
```diff
@@ -477,32 +649,36 @@ def test_args_passed_to_tabulate(self, history):
...
+
         sink = print_log.sink
         row = sink.call_args_list[2][0][0]
         items = row.split()
 
-        assert len(items) == 2  # no valid
-        # epoch
-        assert items[0] == '1'
-        # color 1 used for item 1
-        assert items[1] == list(ansi)[1].value + '0.2500' + ansi.ENDC.value
```
Comment (ottonemo) [inline skorch/tests/test_callbacks.py]: The code is gone now it for the future I would suggest not to test side effects like this but to refactor the coloring function and test it separately in a parametrised test

File: skorch/tests/test_callbacks.py [hunk 2]
```diff
@@ -46,70 +44,286 @@ def test_on_x_methods_have_kwargs(self, callbacks, on_x_methods):
...
+            self, net_cls, module_cls, scoring_cls, train_split, data,
+    ):
+        net = net_cls(
+            module_cls,
+            callbacks=[scoring_cls(scoring=None, name=None)],
+            max_epochs=1,
+            train_split=train_split,
+        )
+        net.fit(*data)
+        assert net.history[:, 'score']
```
[the remark this answers is not in the document]
Comment (benjamin-work) [reply skorch/tests/test_callbacks.py]: `'score'` is consistent with sklearn nomenclature.

**Comment by benjamin-work** [comment]
@ottonemo Please review again.

File: skorch/tests/test_callbacks.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ottonemo) [reply skorch/tests/test_callbacks.py]: Ok then leave it as is

**Pull request merged**